### PR TITLE
CC-374 Support basic auth with secure Elasticsearch

### PR DESCRIPTION
**Pull request by rayokota**
This provides the ability to configure username and password for accessing a secure Elasticsearch cluster.

### CODE REVIEW

File: src/main/java/io/confluent/connect/elasticsearch/ElasticsearchSinkConnectorConfig.java [hunk 1]
```diff
@@ -33,6 +33,16 @@
   private static final String CONNECTION_URL_DOC =
       "List of Elasticsearch HTTP connection URLs e.g. ``http://eshost1:9200,"
       + "http://eshost2:9200``.";
+  public static final String CONNECTION_USERNAME_CONFIG = "connection.username";
+  private static final String CONNECTION_USERNAME_DOC =
+      "The username used to authenticate with Elasticsearch. "
+      + "The default is the empty string, and authentication will only be performed if "
+      + " both the username and password are non-empty.";
```
Comment (rhauch) [inline src/main/java/io/confluent/connect/elasticsearch/ElasticsearchSinkConnectorConfig.java]: We don't want to authenticate when the username is non-null, regardless of the password?

Comment (rayokota) [reply src/main/java/io/confluent/connect/elasticsearch/ElasticsearchSinkConnectorConfig.java]: You mean allow for an empty username or password?  It's a bad security practice.  Do you think some customers will want it?

File: src/main/java/io/confluent/connect/elasticsearch/jest/JestElasticsearchClient.java [hunk 2]
```diff
@@ -108,22 +110,36 @@ public JestElasticsearchClient(String address) {
...
-      JestClientFactory factory = new JestClientFactory();
-      factory.setHttpClientConfig(new HttpClientConfig.Builder(address)
+      HttpClientConfig.Builder builder = new HttpClientConfig.Builder(address)
           .connTimeout(connTimeout)
           .readTimeout(readTimeout)
-          .multiThreaded(true)
-          .build()
-      );
+          .multiThreaded(true);
+      if (username != null && !username.isEmpty() && password != null && !password.isEmpty()) {
```
Comment (mageshn) [inline src/main/java/io/confluent/connect/elasticsearch/jest/JestElasticsearchClient.java]: If we are doing only Basic Auth should we be we using preemptive auth?

Comment (rayokota) [reply src/main/java/io/confluent/connect/elasticsearch/jest/JestElasticsearchClient.java]: Yes, it is using preemptive auth on line 138.

**Comment by rayokota** [comment]
@rhauch , I made a change to allow for username and/or password to be empty

**Comment by rhauch** [comment]
@rayokota please be sure to update #45

File: src/main/java/io/confluent/connect/elasticsearch/ElasticsearchSinkConnectorConfig.java [hunk 3]
```diff
@@ -158,6 +168,26 @@ private static void addConnectorConfigs(ConfigDef configDef) {
...
+        null,
+        Importance.MEDIUM,
+        CONNECTION_USERNAME_DOC,
+        group,
+        ++order,
+        Width.SHORT,
+        "Connection Username"
+    ).define(
+        CONNECTION_PASSWORD_CONFIG,
+        Type.STRING,
```
Comment (kkonstantine) [inline src/main/java/io/confluent/connect/elasticsearch/ElasticsearchSinkConnectorConfig.java]: This needs to change to `Type.PASSWORD` so it won't be leaked in configs.

**Comment by kkonstantine** [comment]
Type for field password is wrong. PR needs to be amended with a follow-up PR.

**Comment by fmeeuw** [comment]
Great that this pr has been merged! I was wondering if its possible to backport this pr (and the other dependent ones) to a version 4.1.x release?

**Comment by rhauch** [comment]
@fmeeuw we generally try to avoid backporting features to earlier versions. Is there a reason you wouldn't want to upgrade to 5.0.0 when it is out? Other than dropping support for JDK 7, this is so far the only change to the ES connector since 4.1.0.

**Comment by cricket007** [comment]
Close #112 ?

**Pull request merged**